### PR TITLE
Defer inference for homomorphic mapped types

### LINKED ISSUE #19951: Stack overflow with recursive mapped types
From https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/lodash

cc @sandersn @ahejlsberg 

**TypeScript Version:**  master

**Code**

```ts
type PartialDeep<T> = {
    [P in keyof T]?: PartialDeep<T[P]>;
};
interface Struct {
    a: number;
    b: string;
    c: boolean;
}

declare function f<T>(): PartialDeep<T>;
var result: Partial<Struct> = f();
```

**Expected behavior:**

pass

**Actual behavior:**

```
$ node built/local/tsc.js index.ts
...\TypeScript\built\local\tsc.js:73240
                throw e;
                ^

RangeError: Maximum call stack size exceeded
    at ...\TypeScript\built\local\tsc.js:31397:29
    at Object.map (...\TypeScript\built\local\tsc.js:1703:29)
    at getAnonymousTypeInstantiation (...\TypeScript\built\local\tsc.js:31498:40)
    at instantiateType (...\TypeScript\built\local\tsc.js:31584:32)
    at getTemplateTypeFromMappedType (...\TypeScript\built\local\tsc.js:29217:21)
    at inferTypeForHomomorphicMappedType (...\TypeScript\built\local\tsc.js:33779:32)
    at inferFromObjectTypes (...\TypeScript\built\local\tsc.js:34031:48)
    at inferFromTypes (...\TypeScript\built\local\tsc.js:33985:29)
    at inferTypes (...\TypeScript\built\local\tsc.js:33824:13)
    at inferTargetType (...\TypeScript\built\local\tsc.js:33803:17)
```

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #19951 in a more scalable way by deferring inference on property that result from inference of homomorphic mapped types. It makes lodash's uses of PartialDeep have very little overhead, since those inferences are lower priority and will never be fully instantiated.

The principle drawback to this is that printing an inferred mapped type is non-trivial. I poked at the problem for a while and didn't arrive at a satisfactory solution. In this PR I just use `{}` for printing the type of all deferred symbols. This is fine for interactive use, OK for our baselines, and not sufficient for packages that want to generate a .d.ts from a file that exports an inferred mapped type. However, I don't think this happens often.

This PR obsoletes #20622.

### CODE REVIEW

**Comment by weswigham** [comment]
> This is fine for interactive use

Do you have any tests for quickinfo for one of these deferred types? If it's the same as the type baselines, I'm not sure it'd be a great experience - minimally it'd be a degradation over our current experience, which is never good.

**Comment by sandersn** [comment]
@ahejlsberg I think this is ready for you to look at.

**Comment by sandersn** [comment]
@weswigham I added quickinfo and completion tests. I didn't add signature info tests since deferred inferred types do not have signatures, at least until we decide to try to infer those too. Given that limitation, the editing experience is .. OK. Completions are fine but quick info just displays `{}` for the type.

I'm going to try putting in a depth cutoff to the printing, which will cover the common, non-recursive case of depth=1 better.

**Comment by sandersn** [comment]
@weswigham After playing with a cutoff, and then discussion with @mhegazy, I decided to keep the current solution, but emit `any` instead of `{}`, since that will never break d.ts, and is slightly less misleading in interactive use.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -11275,42 +11305,41 @@ namespace ts {
          * property is computed by inferring from the source property type to X for the type
          * variable T[P] (i.e. we treat the type T[P] as the type variable we're inferring for).
          */
-        function inferTypeForHomomorphicMappedType(source: Type, target: MappedType, mappedTypeStack: string[]): Type {
+        function inferTypeForHomomorphicMappedType(source: Type, target: MappedType): Type {
+            const key = source.id + "," + target.id;
+            if (deferredInferenceCache.has(key)) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Do we actually need this cache?

File: src/compiler/types.ts [hunk 2]
```diff
@@ -3601,6 +3609,12 @@ namespace ts {
         finalArrayType?: Type;  // Final array type of evolving array type
     }
 
+    /* @internal */
+    export interface DeferredMappedType extends ObjectType {
```
Comment (ahejlsberg) [inline src/compiler/types.ts]: I would call this `ReverseMappedType`.

File: src/compiler/types.ts [hunk 3]
```diff
@@ -3494,6 +3501,7 @@ namespace ts {
         EvolvingArray    = 1 << 8,  // Evolving array type
         ObjectLiteralPatternWithComputedProperties = 1 << 9,  // Object literal pattern with computed properties
         ContainsSpread   = 1 << 10, // Object literal contains spread operation
+        Deferred = 1 << 11, // Object contains a deferred inferred property
```
Comment (ahejlsberg) [inline src/compiler/types.ts]: `ReverseMapped`?

File: src/compiler/types.ts [hunk 4]
```diff
@@ -3260,6 +3261,12 @@ namespace ts {
         isRestParameter?: boolean;
     }
 
+    /* @internal */
+    export interface DeferredTransientSymbol extends TransientSymbol {
```
Comment (ahejlsberg) [inline src/compiler/types.ts]: `ReverseMappedSymbol`?

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -11275,42 +11305,41 @@ namespace ts {
...
+            deferred.mappedType = target;
+            return deferred;
+        }
 
-            function inferTargetType(sourceType: Type): Type {
-                inference.candidates = undefined;
-                inferTypes(inferences, sourceType, templateType, 0, mappedTypeStack);
-                return inference.candidates ? getUnionType(inference.candidates, UnionReduction.Subtype) : emptyObjectType;
-            }
+        function inferDeferredMappedType(sourceType: Type, target: MappedType): Type {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: `getTypeOfReverseMappedSymbol` and pass in the symbol.

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/checker.ts]: Yep, turns out some of the tests run out of memory without it.

File: src/compiler/checker.ts (hunk 5 above)
Comment (sandersn) [reply src/compiler/checker.ts]: This is called separately when getting the type of the index signature, which doesn't have a symbol. However, I think uniformity is valuable enough here to create a simple function that delegates to the original.

**Pull request merged**